src/Features/LanguageServer/ProtocolUnitTests/CodeActions/CodeActionsTests.cs

```diff
@@ -0,0 +1,92 @@
+        [Fact]
+        public async Task TestGetCodeActionsAsync()
+        {
+            var markup =
+@"class A
+{
+    void M()
+    {
+        {|caret:|}int i = 1;
+    }
+}";
+            var (solution, locations) = CreateTestSolution(markup);
+            var expected = CreateCommand("Use implicit type", locations["caret"].First());
+
+            var results = await RunCodeActionsAsync(solution, locations["caret"].First());
+            AssertCodeActionCommandsEqual(expected, results.Single());
+        }
+
+        private static async Task<LSP.Command[]> RunCodeActionsAsync(Solution solution, LSP.Location caret)
+            => await GetLanguageServer(solution).GetCodeActionsAsync(solution, CreateCodeActionParams(caret), new LSP.ClientCapabilities(), CancellationToken.None);
+
+        private static void AssertCodeActionCommandsEqual(LSP.Command expected, LSP.Command actual)
+        {
+            Assert.Equal(expected.Title, actual.Title);
+            Assert.Equal(expected.CommandIdentifier, actual.CommandIdentifier);
+            Assert.Equal(expected.Arguments.Length, actual.Arguments.Length);
+
+            for (var i = 0; i < expected.Arguments.Length; i++)
+            {
+                if (expected.Arguments[i] is LSP.Command)
+                {
+                    AssertCodeActionCommandsEqual((LSP.Command)expected.Arguments[i], (LSP.Command)actual.Arguments[i]);
+                }
+                else if (expected.Arguments[i] is RunCodeActionParams)
+                {
+                    AssertRunCodeActionParamsEqual((RunCodeActionParams)expected.Arguments[i], (RunCodeActionParams)actual.Arguments[i]);
+                }
+            }
+
+            // local functions
+            static void AssertRunCodeActionParamsEqual(RunCodeActionParams expected, RunCodeActionParams actual)
+            {
+                Assert.Equal(expected.Title, actual.Title);
+                Assert.Equal(expected.TextDocument.Uri, actual.TextDocument.Uri);
+                Assert.Equal(expected.Range, actual.Range);
+            }
+        }
+
+        private static LSP.CodeActionParams CreateCodeActionParams(LSP.Location caret)
+            => new LSP.CodeActionParams()
+            {
+                TextDocument = CreateTextDocumentIdentifier(caret.Uri),
+                Range = caret.Range,
+                Context = new LSP.CodeActionContext()
+                {
+                    // TODO - Code actions should respect context.
+                }
+            };
+
+        private static LSP.Command CreateCommand(string title, LSP.Location location)
+            => new LSP.Command()
+            {
+                Title = title,
+                CommandIdentifier = "_liveshare.remotecommand.Roslyn",
+                Arguments = new object[]
+                {
+                    new LSP.Command()
+                    {
+                        Title = title,
+                        CommandIdentifier = "Roslyn.RunCodeAction",
```

Comment:
what does this identifier mean?

Reply:
in https://microsoft.github.io/language-server-protocol/specification (search for the first usage of command, I can't link to it for some reason).

It represents the command handler since the command can basically be anything.

Reply:
so this is sepcific to live share commend then.